documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+# Array Expression Lowering
+This document describes how Fortran array expressions will be lowered from their `Fortran::evaluate::Expr` tree representation to a sequence of FIR operations.
+Some higher-level design regarding array expression evaluation is presented in [Array Expression Evaluation Design Document](documentation/ArrayComposition.md).
+The objectives of array expression lowering to FIR are to produce a sequence of FIR operations that:
+1.  Implements Fortran Semantics and f18 semantic extensions
+2.  Exposes enough structure so that array optimizations can later be performed on FIR
+3.  Either implements, or makes it possible to implement as FIR transformations the expressions rewriting exposed in [Array Expression Evaluation Design Document](documentation/ArrayComposition.md).
+## Discussions about the objectives
```

Comment:
Are there two potentially separable parts to this lowering process?

I'm wondering if there could (and should) be a first part that is the official lowering from the front end to FIR that leaves arrays, array sections, and scalars as arrays, array sections, and scalars, possibly emboxing whichever of those require that in order to be represented in FIR.  If necessary, any object that requires emboxing could be repeatedly emboxed as often as needed to satisfy correctness requirements, possibly on the expectation that one or more downstream CSE passes could eliminate some of those as applicable.

Once valid "high-level" FIR is generated, array-aware optimization can be performed, with lowering to fir loops done somewhere downstream, possibly as late as during generation of LLVM IR.  If there are two (or more) choices for doing this second lowering step, there is a possibility of doing both, as determined on a case by case basis.

Reply:
For array sections, that is more or less what is happening, sub-expressions that are array sections are emboxed (*). However, in general FIR does not provide elemental operations, so we must lower to `fir.loop` during the lowering (for instance,  `fir.fadd` for floating-point add expects only scalar so we cannot lower delay lowering of `A+B` to a `fir.loop` in FIR).

(*) There is a discussion below that current `fir.box` being based on memory reference, it is generally unsafe to leave array section as box if it is the expression top level because it is hard to guarantee the memory reference will not been modified when we propagate the box outside of the statement.